### DISCUSSION TITLE
[@types/prettier] Breaking Change in  (#59690)

### DISCUSSION THREAD

**Discussion by corymhall** [DefinitelyTyped/DefinitelyTyped#60310] [Issues with a @types package]
PR #59690 increased the minimum typescript version from `3.7` to `4.2`, but was released as a `patch` release. This should have been released as a major version bump since it is breaking for anybody still using typescript `^3`.

Can this change be reverted?

[a comment by an automated account is not shown]

**Comment by biffgaut** [comment]
At the least, the previous version should be re-released as the next patch ASAP.

**Comment by JounQin** [comment]
I think you should just upgrade TypeScript or locking `@types/prettier`. The version strategy is related to `prettier`'s version only.

**Reply by corymhall** [reply]
> I think you should just upgrade TypeScript or locking `@types/prettier`. The version strategy is related to `prettier`'s version only.

Shouldn't the versioning strategy of a _types_ package also adhere to the versioning strategy of (type)Script?  Sometimes it is not easy to upgrade major versions (typescript 3.9 -> 4.2) due to breaking changes.

**Reply by mike-lischke** [reply]
Not only that, but what if you write the first typings for a project which is already at version 10? Would you also set 10 for this very first typings package? That would make absolutely no sense. There should be 2 independent versions, one for the base project and one for the typings.

**Reply by JounQin** [reply]
It is possible to support multiple typescript versions for `types` packages at one version, see https://github.com/DefinitelyTyped/DefinitelyTyped/pull/59690#issuecomment-1125576421

If you want, I think you should just draft a PR.

**Reply by mike-lischke** [reply]
It should be clear that my comments don't specifically target typescript differences, but are more of a general nature. It's not relevant in this context that you can define things for different typescript versions. It's about semantic versioning and how you deal with changes. And breaking changes never must go into a patch version.

**Reply by nanoTitan** [reply]
I tried upgrading typescript to latest 2.7.4, and that didn't help. The only way I was able to get this working again was to lock prettier to dev dependencies @2.6.0

**Reply by Shinigami92** [reply]
TypeScript latest is 4.7.2 👀 
TypeScript v2 is fully unsupported!

**Reply by BendingBender** [reply]
This is not how versioning in DefinitelyTyped works. The only version that is internal to a DT package is the patch version of a package. The major and minor versions __always__ adhere to the versions of the package being typed. DefinitelyTyped publishes tags for every package that is released (look here: https://www.npmjs.com/package/@types/prettier?activeTab=versions), you may use one of the tags that corresponds to your typescript version.

Generally speaking, types packages tend to always being improved in breaking ways, every change that improves types to be more concrete or to use a newly released feature from a new compiler version would constitute a breaking change in normal circumstances. But this is actually completely irrelevant for DT packages, they don't work like this. It's comparable to TypeScript compiler versions, here, pretty much every minor release contains breaking changes. You just need to be aware of it.

The only way to enforce strict semver rules for a package is to include the type definitions with the package and write tests that make sure that incompatible changes are detected. You're free to do the necessary persuading of package owners and the work to actually contribute the types and tests for the packages you care about.

**Comment by JounQin** [comment]
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/59690#issuecomment-1125576421

It should be able to change without breaking low level ts users at the same time.

**Comment by AlwinGarside** [comment]
The issue is there is no typescript dependency in the package.json. If there were, NPM would refrain from updating `@types/prettier` to a version that's incompatible with the TypeScript version that's installed.

**Comment by eggsys** [comment]
Which version of this @types/prettier got this breaking change ?
will version fixing solve this problem ?
 
 for example if we fix  Minor version
@types/prettier : [version that not break]

**Reply by MaikLangeDeloitte** [reply]
@types/prettier version 2.6.1 has the breaking change. 

Fixing the version to @types/prettier 2.6.0 can solve the issue for you if you use @types/prettier directly and non of your other dependencies uses @types/prettier without a fixed version. For example if you use jest you can still run into this issue, as jest-snapshot pulls the current @types/prettier version.

**Reply by eggsys** [reply]
ok 
there is something i didn't try yet 
can we fixing all of @types/prettier version of dependencies that use @types/prettier in package-lock.json
then we run `npm ci` instend of   `npm install` 

so its should not upgrade any version of @types/prettie in Dependecies

**Reply by NoahGrayDL** [reply]
For me, @types/prettier 2.6.0 also returns the similar error. So I downgrade the @types/prettier version as 2.4.4 in the package-lock.json because my another ts project has that version and working well... 
+ Don't forget to update the tgz link and jest-snapshot's prettier ^2.0.0 also in the package-lock.json 😂 I missed them at the first time...

**Reply by gburke-ppb** [reply]
Adding `"@types/prettier": "<2.6.0",` to my package.json fixed the issue for me.
`jest` is pulling in `prettier` -- I didn't previously have to add it to my package.json.
Very annoying that I have to do this because someone decided on a breaking change in a point release.

**Reply by bbrenceOLabs** [reply]
@types/prettier 2.6.0 also didn't work with my repo. I also ended up upgrading to 4.2.X which was an option luckily. Surprised this was and still is an issue after a month on such a popular library :(

**Reply by tannerdolby** [reply]
With TypeScript v4.9.5 and Jest v29.4.2, I saw weird type errors related to the prettier types when attempting to compile a file with `tsc foo.ts`. Using `npx tsc foo.ts` allowed the file to be compiled as expected.

**Reply by fyzhu** [reply]
> For me, @types/prettier 2.6.0 also returns the similar error. So I downgrade the @types/prettier version as 2.4.4 in the package-lock.json because my another ts project has that version and working well...
> 
> * Don't forget to update the tgz link and jest-snapshot's prettier ^2.0.0 also in the package-lock.json 😂 I missed them at the first time...

@types/prettier 2.4.4 work for me

**Comment by TylerSustare** [comment]
This was a bummer. Upgrading TS did work for me but would have been much riskier in a giant codebase.

**Reply by elopezanaya** [reply]
to what version did you upgrade TS?
